### PR TITLE
[IMP] pre-commmit: Run prettier and eslint as local hooks and update versions in v18

### PR DESCRIPTION
**Pull request by PabloEForgeFlow**
Should fix https://github.com/OCA/oca-addons-repo-template/issues/264 and https://github.com/OCA/oca-addons-repo-template/issues/224.

### CODE REVIEW

**Comment by PabloEForgeFlow** [comment]
@sbidoul I made the `prettier` and `eslint` hooks local and updated versions at the same time. Perhaps they should be two separate PRs?

Also, pre-commit was [failing](https://github.com/OCA/oca-addons-repo-template/actions/runs/11249365414/job/31276135881), due to the `.prettierrc.yml` symlink being broken, so I updated it to the new `'{% if 12 < odoo_version < 18 %}.prettierrc.yml{% endif %}'`, but we'll probably want to use the v18 file once we're sure it's working.

**Comment by sbidoul** [comment]
Is it not necessary to convert eslintrc.yml to cjs, and to the new eslint 9 config format for v18?

We could also upgrade node for v18.

BTW, I wonder why GitHub is still trying to run the tests for 11 and 12.

**Comment by PabloEForgeFlow** [comment]
> Is it not necessary to convert eslintrc.yml to cjs, and to the new eslint 9 config format for v18?

Will do that, I thought it was something limited to prettier, but it does appear to have [changed as well](https://eslint.org/docs/latest/use/configure/configuration-files). Guess that's the new way of doing configuration files in the javascript world.

> We could also upgrade node for v18.

I will look into it, see what version makes most sense.

> BTW, I wonder why GitHub is still trying to run the tests for 11 and 12.

Yeah that's weird, maybe something's still cached on GitHub's side?

**Comment by PabloEForgeFlow** [comment]
I've migrated the eslint configuration to new format following both [eslint](https://eslint.org/docs/latest/use/configure/migration-guide) and [jsdoc](https://github.com/gajus/eslint-plugin-jsdoc/wiki/Comparison-with-deprecated-JSdoc-related-ESLint-rules) guides. Most parts have been migrated using the [configuration migrator](https://www.npmjs.com/package/@eslint/migrate-config), but some, such as jsdoc settings, were done by hand. I also set `ecmaVersion` to 2024, but I'm not sure if we would rather stay on 2022.

I also update the node to version 22, which is the latest TLS release, and fixed some issues with prettier.

**Comment by sbidoul** [comment]
Hi @PabloEForgeFlow I did a test by applying the template from your PR onto https://github.com/OCA/storage/pull/399

I noticed two things:

1. In 18+, we need to remove `.prettierrc.yml` otherwise it takes precedence over the cjs config. We also need to remove the old `.eslintrc.yml`. This can be done in `copier.yml`, `_tasks`.
2. I think we need to exclude `^eslint.config.cjs|^prettier.config.cjs`, otherwise eslint is trying to check them and is not happy with what it sees.

**Comment by PabloEForgeFlow** [comment]
Makes sense, I added a line to `_tasks` removing them and put both in the global pre-commit exclude.

File: copier.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ _min_copier_version: "9"
 
 _tasks:
   - rm -f {% if ci != "Travis" %}.travis.yml{% endif %} .t2d.yml
+  - rm -f .prettierrc.yml .eslintrc.yml
```
Comment (sbidoul) [inline copier.yml:8]: @PabloEForgeFlow these should be removed only for odoo_version >= 18, right? Can you adjust that?

Comment (sbidoul) [reply copier.yml:8]: Done in 6f0c78888e983606697d4d4c8cf52bf4c3eec471

Comment (PabloEForgeFlow) [reply copier.yml:8]: I though they would be recreated, but yeah it's probably more clear this way.

**Pull request merged**